packages/powersync_attachments_stream/lib/src/attachment_queue_service.dart

```diff
@@ -0,0 +1,404 @@
+            .firstWhere(
+              (a) => a != null && a.id == item.id,
+              orElse: () => null,
+            );
+
+        if (existingQueueItem == null) {
+          if (!downloadAttachments) continue;
+
+          // This item should be added to the queue.
+          // This item is assumed to be coming from an upstream sync.
+          final String filename =
+              item.filename ??
+              await resolveNewAttachmentFilename(item.id, item.fileExtension);
+
+          attachmentUpdates.add(
+            Attachment(
+              id: item.id,
+              filename: filename,
+              state: AttachmentState.queuedDownload,
+              metaData: item.metaData,
+              // hasSynced: existingQueueItem.hasSynced;
+            ),
+          );
+        } else if (existingQueueItem.state == AttachmentState.archived) {
+          // The attachment is present again. Need to queue it for sync.
+          if (existingQueueItem.hasSynced) {
+            // No remote action required, we can restore the record (avoids deletion).
+            attachmentUpdates.add(
+              existingQueueItem.copyWith(state: AttachmentState.synced),
+            );
+          } else {
+            // The localURI should be set if the record was meant to be downloaded
+            // and has been synced. If it's missing and hasSynced is false then
+            // it must be an upload operation.
+            attachmentUpdates.add(
+              existingQueueItem.copyWith(
+                state: existingQueueItem.localUri == null
+                    ? AttachmentState.queuedDownload
+                    : AttachmentState.queuedUpload,
+              ),
+            );
+          }
+        }
+      }
+
+      // Archive any items not specified in the watched items.
+      // For QUEUED_DELETE or QUEUED_UPLOAD states, archive only if hasSynced is true.
+      // For other states, archive if the record is not found in the items.
+      for (final attachment in currentAttachments) {
+        final notInWatchedItems = items.every(
+          (update) => update.id != attachment.id,
+        );
+
+        if (notInWatchedItems) {
+          switch (attachment.state) {
+            case AttachmentState.queuedDelete:
+            case AttachmentState.queuedUpload:
+              if (attachment.hasSynced) {
+                attachmentUpdates.add(
+                  attachment.copyWith(state: AttachmentState.archived),
+                );
+              }
+              break;
+            default:
+              attachmentUpdates.add(
+                attachment.copyWith(state: AttachmentState.archived),
+              );
+          }
+        }
+      }
+
+      await context.saveAttachments(attachmentUpdates);
+    });
+  }
+
+  /// Creates a new attachment locally and queues it for upload.
+  /// The filename is resolved using [resolveNewAttachmentFilename].
+  Future<Attachment> saveFile({
+    required Stream<Uint8List> data,
```

Comment:
I noticed we had to cast in the demo to get a `Stream<Uint8List>`, so I wonder if we should take a `List<int>` as an input here instead (when we expose assets as bytes we should definitely use `Stream<Uint8List>` for that, but it looks like for local writes we can get away with `List<int>` too).